src/jit/decomposelongs.cpp

```diff
@@ -964,20 +975,252 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
             unreached();
     }
 
-    GenTreeArgList* argList = m_compiler->gtNewArgList(loOp1, hiOp1, shiftWidthOp);
+    // If we are shifting by a constant int, we do not want to save to a helper, instead, we decompose.
+    if (shiftOper == GT_CNS_INT)
+    {
+        GenTree* loCopy  = m_compiler->gtNewLclvNode(loOp1LclNum, TYP_INT);
 
-    GenTree* call = m_compiler->gtNewHelperCallNode(helper, TYP_LONG, 0, argList);
+        GenTree* loResult;
+        GenTree* hiResult;
 
-    GenTreeCall*    callNode    = call->AsCall();
-    ReturnTypeDesc* retTypeDesc = callNode->GetReturnTypeDesc();
-    retTypeDesc->InitializeLongReturnType(m_compiler);
+        switch (oper)
+        {
+            case GT_LSH:
+            {
+                if (count == 0)
+                {
+                    // Do nothing.
+                    loResult = loOp1;
+                    hiResult = hiOp1;
+                    Range().InsertAfter(tree, loResult, hiResult);
+                }
+                else if (count < 32)
+                {
+                    // Hi is a GT_LSH_HI, lo is a GT_LSH. Will produce:
+                    // reg1 = lo
+                    // shl lo, shift
+                    // shld hi, reg1, shift
+
```

Comment:
By adding an insertAfter argument, you mean changing the signature to:

`GenTree* DecomposeLongs::FinalizeDecomposition(LIR::Use& use, GenTree* loResult, GenTree* hiResult, GenTree* insertAfter)`

And most calls to it would be changed to:

`FinalizeDecomposition(use, loResult, hiResult, hiResult)`

Except in the LSH case, where we'd call:

`FinalizeDecomposition(use, loResult, hiResult, loResult)`